Added accuracy as attribute to basis

nikwitt opened the pull request:
Added relative tolerance `eps > s_lmax / s_0` as an attribute to the basis class (and inheriting it in basis_set). It is a solution to issue #19, although the way of adding a value for `eps` is kinda hacky if `eps=None` is the input. Like this, it should be more or less consistent (instead of just using `eps = s[-1] / s[0]`, since it can lead to larger basis sizes using this calculated eps).
In the future, adding some SVE_results class might be useful for doing this consistently.

In adding the tolerance, I also cleaned up the attributes/properties of IRbasis and FiniteTempBasis.

nikwitt committed:
Added eps as attribute to basis

shinaoka commented:
Thank you!
But, some test fails, which is not your fault.
In the current implementation, the `sve_result` attribute of a `FiniteTempBasis` object returns SVE results in the tau convention ([0, beta]).
This is not what we expect in creating a new basis object with the same `sve_result`, which must be in the reduced variable `x` ([-1, 1]).
Can you give me a write permission to your repository? I will fix it.

nikwitt commented:
That is interesting. I had this issue when testing as well, but somehow I managed to get it to work (otherwise I would not have submitted). You should be able to fix it now.

shinaoka committed:
Define sve_result attribute of FiniteTempBasis in the reduced variable

shinaoka commented:
OK, all the tests are now passed.
I am still trying to understand the logic for setting the value of the attribute `eps` when `eps=None` is given.

My test code:
```
def test_reuse_sve_result():
    beta = 1.
    wmax = 1e+1
    basis = FiniteTempBasis("F", beta, wmax)
    basis2 = FiniteTempBasis("F", beta*2, wmax/2)
    print(basis.eps, basis.s[-1]/basis.s[0])
    assert basis.eps == basis2.eps
```

The results:
```
Precomputing SVEs for logistic kernel ...
2e-15 1.903279158269547e-15
```

The value of `eps` is slightly larger than `s[-1]/s[0]`. I expect the opposite. Is this the intended behavior?

nikwitt commented:
First of all, you are correct that it should be `np.floor` instead of `np.ceil`, i.e., a little bit smaller than larger.

Second, I tested this a bit more now. Although `numpy.nextafter` seems simpler, it does not fully get rid of the problem I had in mind. I admit, the problem I am going to explain is a bit artificial and unlikely to happen, but if it did occur I believe it would not be easy to find.
My reasoning for the logic of setting `eps < eps_s = s[-1] / s[0]` is that when recalculating the basis function by passing `eps_s` _without also passing the sve_result_ the basis size _can_ change. This will most probably lead to some error due to size mismatch between calculations. You can check this behavior using:
```
beta = 1.
wmax = 1e+1
for eps in [None, 1e-8, 1e-7, 1e-5]:
    basis = sparse_ir_test.FiniteTempBasis("F", beta, wmax, eps=eps)
    # reuse the same eps (no problem)
    basis_eps    = sparse_ir_test.FiniteTempBasis("F", beta, wmax, eps=basis.eps)
    # use stricht eps = s_max / s0 <-- this is esp. bad
    basis_sv_eps = sparse_ir_test.FiniteTempBasis("F", beta, wmax, eps=basis.s[-1]/basis.s[0])
    # use next float below s_max / s0 <-- better but sometimes bad
    basis_np_eps = sparse_ir_test.FiniteTempBasis("F", beta, wmax, eps=np.nextafter(basis.s[-1] / basis.s[0],0))

    # create new basis from sve_results
    basis_2 = sparse_ir_test.FiniteTempBasis("F", beta*2, wmax/2, sve_result=basis.sve_result)
    # reuse eps; mind that basis_2.eps != basis.eps, as we floor at a certain precision 
    # (NOT machine precision as np.nextafter)
    basis_2_eps = sparse_ir_test.FiniteTempBasis("F", beta, wmax, eps=basis_2.eps)

    print(eps, basis.eps, basis_eps.eps, basis_sv_eps.eps, basis_np_eps.eps, basis_2.eps, basis_2_eps.eps)
    print(basis.size, basis_eps.size, basis_sv_eps.size, basis_np_eps.size, basis_2.size, basis_2_eps.size)
```
Mind, it does not occur for every eps (that's why I let it iterate over three values). I am not sure of the reason for the different sizes, but I guess the strict condition of `s_max > eps * s_0` instead of `s_max >= eps * s_0` is enforced in the truncation process. If we do not want to change that, setting `eps` slightly smaller than `s[-1] / s[0]` is sufficient to circumvent this. But "slightly" as in "machine precision" from `numpy.nextafter` is not enough as this example demonstrates.

nikwitt committed:
corrected eps ceil to floor

shinaoka commented:
OK, I got the point. 
Minimal code

```
def test_niklas():
    beta = 1.
    wmax = 1e+1
    eps = 1e-8

    # SVE is done in ddouble because eps is small enough
    basis = FiniteTempBasis("F", beta, wmax, eps=eps)

    # SVE is done in double because eps = 6.514192910092466e-08
    basis_np_eps = FiniteTempBasis("F", beta, wmax, eps=np.nextafter(basis.s[-1] / basis.s[0],0))
    assert basis.size == basis_np_eps.size
```

The first SVE is done in `ddouble` (extended precision) but the second one is done in `double`.
So, the computed singular values do not match between the first and second SVE's within the machine precision!

mwallerb commented:
Ouch, yes that's annoying.

There is not really a way of fixing this without storing the discretization parameters together with the SVE.
As a workaround, I suggest **removing** `eps` from `FiniteTempBasisSet` and instead add the following
property to `IRBasis`, `FiniteTempBasis`, and `FiniteTempBasisSet`:

    @property
    def accuracy(self):
         return np.nextafter(basis.s[-1] / basis.s[0], 0)

nikwitt commented:
I agree, for full integrity it probably should be stored together with the SVE in the future.

For now, how about setting both, the accuracy property from `s[-1]/s[0]` and `eps`? But by explicitly mentioning that the `basis.eps` property/attribute is the same as input, not necessarily the same as used in the SVE. Then, if someone did not set eps, it is left as `None`, while others can store, access, and possibly reuse their preset values.
This could still lead to inconsistencies between calculations if someone really wants it to be

shinaoka commented:
This problem seems to be fundamental.
It's all about how to uniquely (and in a numerically robust way) identify a basis.
Ideally, we should use (Λ, size) rather than (Λ, eps).
In practice, we can use (Λ, eps, size) and use eps as a hint for determining work_dtype for SVE.
What do you think, Markus?

mwallerb commented:
I am against `eps` in this form.  `eps` is not useful to reproduce the basis, (you need `work_dtype`, basis size,  potentially gauss order and so on), so it is not useful to store it with the basis on its own.  Moreover, its inconsistent between `FiniteTempBasis` and `FiniteTempBasisSet`.  So, its neither useful in reproducing the basis nor useful in interpreting the basis, so let's not add it.

nikwitt commented:
Okay, I can totally understand that point. Then I will change the `eps` to `accuracy` (does not hurt to have). Implementing a structure `SVEResult` should be dealt with in another commit/PR.

nikwitt committed:
Added accuracy, removed eps attr.

nikwitt commented:
I added the accuracy property now. As it should not be used to reproduce a basis, I think it is sufficient to not use `numpy.nextafter`. To reflect the change, I also changed the title of the PR.
If everything is fine for you, we can close this PR(?).

mwallerb reviewed (approved):
Looks good!

mwallerb commented:
Dear @nikwitt,

unfortunately, several tests fail with your PR, which we only found out after merging, so I had to revert your merge.

Can you open another PR and please make sure that the test suite runs cleanly (`pytest`)?

Thanks!
